libs/sdk-ui-dashboard/src/model/state/_infra/queryProcessing.ts

```diff
@@ -0,0 +1,184 @@
+
+/**
+ * @internal
+ */
+export type QueryEnvelope = {
+    readonly type: typeof QueryEnvelopeActionTypeName;
+    readonly query: IDashboardQuery;
+    readonly onStart: (query: any) => void;
+    readonly onSuccess: (result: any) => void;
+    readonly onError: (err: Error) => void;
+};
+
+function* processQuery(
+    service: IDashboardQueryService<any, any>,
+    ctx: DashboardContext,
+    envelope: QueryEnvelope,
+) {
+    const {
+        query,
+        query: { type, correlationId },
+    } = envelope;
+    const correlationIdForLog = correlationId ?? "(no correlationId provided)";
+
+    try {
+        try {
+            envelope.onStart(query);
+        } catch (e) {
+            // eslint-disable-next-line no-console
+            console.warn(
+                `An error has occurred while calling onStart function provided for ${type}@${correlationIdForLog} processing:`,
+                e,
+            );
+        }
+
+        yield dispatchDashboardEvent(queryStarted(ctx, correlationId));
+
+        const result: SagaIterator<typeof service.generator> = yield call(
+            service.generator,
+            ctx,
+            envelope.query,
+        );
+
+        try {
+            envelope.onSuccess(result);
+        } catch (e) {
+            // eslint-disable-next-line no-console
+            console.warn(
+                `An error has occurred while calling onSuccess function provided for ${type}@${correlationIdForLog} processing`,
+                e,
+            );
+        }
+        yield dispatchDashboardEvent(queryCompleted(ctx, query, result, correlationId));
+    } catch (e) {
+        try {
+            envelope.onError(e);
+        } catch (ne) {
+            // eslint-disable-next-line no-console
+            console.warn(
+                `An error has occurred while calling onError function provided for ${type}@${correlationIdForLog} processing:`,
+                ne,
+            );
+        }
+
+        if (isDashboardQueryFailed(e)) {
+            yield dispatchDashboardEvent(e);
+        } else {
+            yield dispatchDashboardEvent(
+                internalQueryErrorOccurred(
+                    ctx,
+                    `An internal error has occurred while processing ${type}`,
+                    e,
+                    correlationId,
+                ),
+            );
+        }
+    }
+}
+
+function ensureQueryWrappedInEnvelope(action: Action): QueryEnvelope {
+    if (action.type === QueryEnvelopeActionTypeName) {
+        return action as QueryEnvelope;
+    }
+
+    return {
+        type: QueryEnvelopeActionTypeName,
+        query: action,
+        onStart: noop,
+        onSuccess: noop,
+        onError: noop,
+    };
+}
+
+/**
+ * Creates components that should be integrated into the dashboard store in order to facilitate query processing.
+ *
+ * @param queryServices - query services use to initialize the components
+ */
+export function createQueryProcessingComponents(
```

Comment:
I am concerned that using Component for anything else than React Components might cause unnecessary confusion. What about Module or StoreModule?

Reply:
good point, i'll think of something different

Reply:
fixed